### PR TITLE
Apply/Copy preferred entity and Butterbar for Entity list page

### PR DESCRIPTION
**Pull request by RajeshwariJadhav**
# Apply/Copy preferred entity

- **Message passing** enables the DOM's access to the _data id_ of the selected entity. The message is sent by **preferred.js** to **OUSelectContent.js**.
- The Apply/Copy change their name and functionality dynamically depending on the entity type chosen by the admin in **preferred.js.** 
- On the OU selection page, there is an **Apply** button that performs a click on the selected OU button to render its settings page. 
- On the group/user selection page, there is a **Copy** button that copies the name of the selected group/user for the admin to paste into the search bar. 

# Butterbar for improved UX

- The butterbar appears at the top of every entity list page, and has a **Done** button on it. It's implemented in **OUListContent.js** and works for all entities (user, group and OU list).
- When the user initially clicks on the Add/Remove button, the remaining buttons on each entity row are blocked, and only the +/- buttons are visible for clarity.
- When the **Done** button is clicked, the previous buttons (such as rename entity, add another entity) re-appear with full functionality.
- Additionally, all the +/- buttons are removed, and the butterbar itself also disappears.
- The **Done** button thus serves as a way for the user to restore the DOM to exactly how it was before the extension was being used. 
- As the butterbar also has instructions for the user, the extension becomes easier to use, improving UX.

Screen Shot 2020-07-17 at 5 40 42 PM

# Testing

- The _addButterbar_ function in **OUListContent.js** was tested to ensure that the butterbar was appearing in the right place on the entity list page.
- The _restoreDOM_ function in **OUListContent.js** was tested to make sure the DOM was restored with the **Done** button was clicked. Mock storage was used to facilitate _addButtons_ before this function could be called. 

These are the tests in **OUListContent.js**
Screen Shot 2020-07-17 at 5 02 58 PM

### CODE REVIEW

File: STEP_Project/OUListContent.js [hunk 1]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+    }
+    
+    let tabl = document.querySelector('table[role=grid]');
+    tabl.removeEventListener('click',addRemoveButtonClick);
+    let entities = tabl.rows;
+    let numEntities = entities.length;
+    for (let i = 1; i < numEntities; i++) 
+    {
+        let row = entities[i];
+        var toDeleteButton = row.querySelectorAll('td[class$="Class"]')[0];
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:28]: Use `const`

File: STEP_Project/OUListContent.js [hunk 2]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+            toDeleteButton.remove();
+        }
+        if (row.lastChild !== null)
+        {
+            row.lastChild.style.opacity = restoreOpacity;
+            row.lastChild.style.pointerEvents = restoreEvents;
+        }
+    }
+
+    document.getElementsByClassName("butterBar")[0].remove();
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:40]: Use single quotes

File: STEP_Project/OUListContent.js [hunk 3]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+}
+
+/**
+ * Adds a butterbar for better UX.
+ * Contains a DONE button that restores DOM.
+ */
+function addButterbar()
+{
+    let butterBar = document.createElement("div");
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:49]: const butterbar

File: STEP_Project/OUListContent.js [hunk 4]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+    butterBar.setAttribute("class", "butterBar");
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:50]: `butterbar.classList.add('butterBar');`

- Single quotes
- Use specific apis

File: STEP_Project/OUListContent.js [hunk 5]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+    butterBar.style.backgroundColor = "#e8f0fe";
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:51]: Use single quotes for all styles

File: STEP_Project/OUListContent.js [hunk 6]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+    butterBar.style.minHeight = "45px";
+    butterBar.style.borderBottom = "1px solid #e0e0e0";
+    butterBar.style.width = "100%";
+    let cwiz = document.getElementsByTagName("c-wiz")[2].firstChild.firstChild.firstChild;
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:55]: Single quotes

File: STEP_Project/OUListContent.js [hunk 7]
```diff
@@ -3,6 +3,70 @@ if (chrome.storage !== undefined)
...
+    cwiz.insertBefore(butterBar, cwiz.lastChild);
+    let doneButton = document.createElement("p");
+    doneButton.innerHTML = "DONE"
+    doneButton.style.float = "right"
+    doneButton.style.paddingRight = "50px"
+    doneButton.style.color = "#3367d6"
+    let instructions = document.createElement("p");
```
Comment (areeb1992) [inline STEP_Project/OUListContent.js:62]: use const

File: STEP_Project/OUSelectContent.js [hunk 8]
```diff
@@ -44,9 +44,27 @@ function selectClick ()
  */
 function onMessageFunction(request)
 {
-    let queryVar = '[data-content-id=\'' + request.dataId + '\']';
-    let orgUnits = document.querySelectorAll(queryVar);
-    orgUnits[0].children[1].firstChild.click();
+    let dataId = request.dataId[0];
```
Comment (areeb1992) [inline STEP_Project/OUSelectContent.js:47]: Use const if variable is not meant to change

**Comment by RajeshwariJadhav** [comment]
All requested changes were made:
- const variables were used wherever possible
- eslint suggestions were fixed for code styling

File: STEP_Project/OUListContent.js [hunk 9]
```diff
@@ -173,7 +237,9 @@ function addButtonsToRows(data)
             //hide unnecessary buttons
             if (row.lastChild !== null)
             {
+                restoreOpacity = row.lastChild.style.opacity;
                 row.lastChild.style.opacity = 0;
+                restoreEvents = row.lastChild.style.pointerEvents;
                 row.lastChild.style.pointerEvents = 'none';
```
Comment (atishey-jain) [inline STEP_Project/OUListContent.js:243]: This piece of code is inside a loop. Changing global variable here means that only last iteration of the loop will be retained in this global variable. Similarly for when we are restoring using this global variable. This brings up some obvious questions, do all rows have the same opacity and pointerEvents ? If no, then this won't work and we need to have an array of these. If yes, then we shouldn't set this inside a for loop and instead set it explicitly to one of the row's attributes outside the loop. Also I don't think maintaining this global state is a very nice approach, we should just pass it around in each function around in each function and in listeners.

Comment (areeb1992) [reply STEP_Project/OUListContent.js:243]: One approach could be, you can also copy the original style into data attribute for each node and restore it later.

`row.lastChild.setAttribute('data-style',  row.getAttribute('style'));` for copying

`row.lastChild.setAttribute('style', row.getAttribute('data-style'));` for restoring

With this approach you won't have to deal with what styles are you changing

Comment (RajeshwariJadhav) [reply STEP_Project/OUListContent.js]: That's a really good idea, I implemented it. 
Now there's no more global variables, and the style is set separately for each row and restored separately for each row. Although all rows have the same style so far, I've maintained the for loop for adaptability.

**Pull request merged**